lib/parser.mly

```diff
@@ -1,599 +0,0 @@
-    obtd=option(transformed_data_block)
-    obp=option(parameters_block)
-    obtp=option(transformed_parameters_block)
-    obm=option(model_block)
-    obg=option(generated_quantities_block)
-    EOF
-    {
-      grammar_logger "program" ;
-      { functionblock= obf
-      ; datablock= obd
-      ; transformeddatablock= obtd
-      ; parametersblock= obp
-      ; transformedparametersblock= obtp
-      ; modelblock= obm
-      ; generatedquantitiesblock= obg }
-    }
-
-(* blocks *)
-function_block:
-  | FUNCTIONBLOCK LBRACE fd=list(function_def) RBRACE
-    {  grammar_logger "function_block" ; fd}
-
-data_block:
-  | DATABLOCK LBRACE tvd=list(top_var_decl_no_assign) RBRACE
-    { grammar_logger "data_block" ; tvd }
-
-transformed_data_block:
-  | TRANSFORMEDDATABLOCK LBRACE tvds=list(top_vardecl_or_statement) RBRACE
-    {  grammar_logger "transformed_data_block" ;  tvds }
-    (* NOTE: We are choosing to allow mixing of statements and top_var_decls *)
-
-parameters_block:
-  | PARAMETERSBLOCK LBRACE tvd=list(top_var_decl_no_assign) RBRACE
-    { grammar_logger "parameters_block" ; tvd }
-
-transformed_parameters_block:
-  | TRANSFORMEDPARAMETERSBLOCK LBRACE tvds=list(top_vardecl_or_statement) RBRACE
-    { grammar_logger "transformed_parameters_block" ; tvds }
-
-model_block:
-  | MODELBLOCK LBRACE vds=list(vardecl_or_statement) RBRACE
-    { grammar_logger "model_block" ; vds  }
-
-generated_quantities_block:
-  | GENERATEDQUANTITIESBLOCK LBRACE tvds=list(top_vardecl_or_statement) RBRACE
-    { grammar_logger "generated_quantities_block" ; tvds }
-
-(* function definitions *)
-identifier:
-  | id=IDENTIFIER
-    {
-      grammar_logger ("identifier " ^ id) ;
-      {name=id; id_loc=Location ($startpos, $endpos)}
-    }
-  | TRUNCATE
-    {
-      grammar_logger "identifier T" ;
-      {name="T"; id_loc=Location ($startpos, $endpos)}
-    }
-
-function_def:
-  | rt=return_type name=identifier LPAREN args=separated_list(COMMA, arg_decl)
-    RPAREN b=statement
-    {
-      grammar_logger "function_def" ;
-      UntypedStmt (FunDef {returntype = rt; funname = name;
-                           arguments = args; body=b;},
-      initialize_stmt_meta $startpos $endpos)
-    }
-
-return_type:
-  | VOID
-    { grammar_logger "return_type VOID" ; Void }
-  | ut=unsized_type
-    {  grammar_logger "return_type unsized_type" ; ReturnType ut }
-
-arg_decl:
-  | od=option(DATABLOCK) ut=unsized_type id=identifier
-    {  grammar_logger "arg_decl" ;
-       match od with None -> (TParam, ut, id) | _ -> (TData, ut, id)  }
-
-unsized_type:
-  | bt=basic_type ud=option(unsized_dims)
-    {  grammar_logger "unsized_type" ;
-       let reparray n x =
-         let rec repeat n f x =
-           if n <= Int64.zero then x else repeat (Int64.pred n) f (f x) in
-         repeat n (fun y -> Array y) x in
-       let size =
-         match ud with Some d -> Int64.succ (Int64.of_int d) | _ -> Int64.zero
-       in
-       reparray size bt    }
```

Comment:
Can we simplify this? Seems like a fold might work.

Reply:
Done.